### PR TITLE
Refactor `afs` and `afsj`

### LINKED ISSUE #2561: afs/afsj return different argument count
Binary: `elf/ls` from https://github.com/rizinorg/rizin-testbins
Cmd: `s 0x13c70;af;afs;afsj`
Output:
```
void fcn.00013c70(int64_t arg1, int64_t arg2, int64_t arg3);
{"name":"fcn.00013c70","noreturn":false,"ret":"void","cc":"amd64","args":[]}
```

### PR DESCRIPTION
**Pull request by theopechli**
**Your checklist for this pull request**
- [X] I've read the [guidelines for contributing](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md) to this repository
- [X] I made sure to follow the project's [coding style](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md#code-style)
- [X] I've documented or updated the documentation of every function and struct this PR changes. If not so I've explained why.
- [ ] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the [rizin book](https://github.com/rizinorg/book) with the relevant information (if needed)

**Detailed description**

This pull request fixes a bug in `afsj`.

I tried to use `rz_type_func_save` inside `rz_analysis_function_derive_type` but it would segfault IIRC, so it is only used when `afsj` is called.

**Test plan**

...

**Closing issues**

closes #2561

### CODE REVIEW

File: librz/analysis/fcn.c [hunk 1]
```diff
@@ -2554,18 +2572,7 @@ RZ_API RZ_OWN RzList /* RzType */ *rz_analysis_types_from_fcn(RzAnalysis *analys
...
- * Checks if the type is defined already for this function, if yes -
- * it returns pointer to the one stored in the types database.
- * If not - it creates a new RzCallable instance based on the function name,
- * its arguments' names and types.
- *
- * \param analysis RzAnalysis instance
- * \param f Function to update
- */
-RZ_API RZ_OWN RzCallable *rz_analysis_function_derive_type(RzAnalysis *analysis, RzAnalysisFunction *f) {
+RZ_API RZ_OWN RzCallable *rz_analysis_function_clone_type(RzAnalysis *analysis, const RzAnalysisFunction *f) {
```
Comment (XVilka) [inline librz/analysis/fcn.c:2584]: Add a brief Doxygen description please

File: librz/analysis/fcn.c [hunk 2]
```diff
@@ -2581,22 +2588,62 @@ RZ_API RZ_OWN RzCallable *rz_analysis_function_derive_type(RzAnalysis *analysis,
...
 	}
-	// If there is no match - create a new one.
+	return NULL;
+}
+
+/**
+ * \brief Creates the RzCallable type for the given function
+ *
+ * Creates the RzCallable type for the given function
+ * by searching in the types database and returns it.
```
Comment (XVilka) [inline librz/analysis/fcn.c]: `and returning it`.

File: librz/analysis/fcn.c [hunk 3]
```diff
@@ -2581,22 +2588,62 @@ RZ_API RZ_OWN RzCallable *rz_analysis_function_derive_type(RzAnalysis *analysis,
...
+ *
+ * \param analysis RzAnalysis instance
+ * \param f Function to update
+ * \param callable A function type
+ */
+RZ_API RZ_OWN RzCallable *rz_analysis_function_create_type(RzAnalysis *analysis, RzAnalysisFunction *f) {
 	// TODO: Figure out if we should use shortname or a fullname here
 	// At this point the `callable` pointer is *owned*
```
Comment (XVilka) [inline librz/analysis/fcn.c]: Can remove this comment then.

File: librz/analysis/fcn.c [hunk 4]
```diff
@@ -2581,22 +2588,62 @@ RZ_API RZ_OWN RzCallable *rz_analysis_function_derive_type(RzAnalysis *analysis,
...
+ *
+ * Gets the given function's arguments (names and types)
+ * and if it has none it simply returns. Otherwise, it
+ * creates RzCallableArgs and adds them to RzCallable.
+ *
+ * \param analysis RzAnalysis instance
+ * \param f Function to update
+ * \param callable A function type
+ */
+RZ_API RZ_OWN bool rz_analysis_function_derive_args(RzAnalysis *analysis, RzAnalysisFunction *f, RzCallable **callable) {
```
Comment (XVilka) [inline librz/analysis/fcn.c]: Same

File: librz/analysis/fcn.c [hunk 5]
```diff
@@ -2581,22 +2588,62 @@ RZ_API RZ_OWN RzCallable *rz_analysis_function_derive_type(RzAnalysis *analysis,
...
+ * \brief Sets the RzCallable return type for the given function
+ *
+ * Checks if the given function's return type exists
+ * and adds it to RzCallable by cloning it.
+ *
+ * \param analysis RzAnalysis instance
+ * \param f Function to update
+ * \param callable A function type
+ */
+RZ_API RZ_OWN void rz_analysis_function_derive_return_type(RzAnalysisFunction *f, RzCallable **callable) {
```
Comment (XVilka) [inline librz/analysis/fcn.c]: RZ_OWN is unnecessary then.

**Review by XVilka — CHANGES_REQUESTED** [review]

File: librz/analysis/fcn.c [hunk 6]
```diff
@@ -2608,17 +2663,49 @@ RZ_API RZ_OWN RzCallable *rz_analysis_function_derive_type(RzAnalysis *analysis,
...
+			return false;
 		}
 		RzCallableArg *arg = rz_type_callable_arg_new(analysis->typedb, var->name, cloned_type);
 		if (!arg) {
 			rz_pvector_free(args);
-			rz_type_callable_free(callable);
-			return NULL;
+			rz_type_callable_free(*callable);
+			RZ_LOG_ERROR("Cannot create callable argument\n");
+			rz_warn_if_reached();
```
Comment (XVilka) [inline librz/analysis/fcn.c]: Remove the `rz_warn_if_reached()`.

File: librz/analysis/fcn.c [hunk 7]
```diff
@@ -2608,17 +2663,49 @@ RZ_API RZ_OWN RzCallable *rz_analysis_function_derive_type(RzAnalysis *analysis,
 		RzType *cloned_type = rz_type_clone(var->type);
 		if (!cloned_type) {
 			rz_pvector_free(args);
-			rz_type_callable_free(callable);
-			return NULL;
+			rz_type_callable_free(*callable);
+			RZ_LOG_ERROR("Cannot parse function's argument type\n");
+			rz_warn_if_reached();
```
Comment (XVilka) [inline librz/analysis/fcn.c]: Remove the `rz_warn_if_reached()`

File: librz/analysis/fcn.c (hunk 3 above)
Comment (XVilka) [reply librz/analysis/fcn.c]: Forgot about this?

**Review by XVilka — APPROVED** [review]
LGTM with a few nitpicks.

File: librz/analysis/fcn.c (hunk 3 above)
Comment (theopechli) [reply librz/analysis/fcn.c]: You want me to remove those 3 comments? Are they not needed?

File: librz/analysis/fcn.c (hunk 7 above)
Comment (ret2libc) [reply librz/analysis/fcn.c]: these are not resolved

File: librz/analysis/fcn.c (hunk 3 above)
Comment (XVilka) [reply librz/analysis/fcn.c]: It was needed when it was a part of the function, now it's always owned. Only two lines, keep the TODO

File: librz/analysis/fcn.c (hunk 7 above)
Comment (theopechli) [reply librz/analysis/fcn.c]: yea i just pushed, sorry

**Review by XVilka — APPROVED** [review]

**Pull request merged**